### PR TITLE
Fix sync on index_select decomposition when the index has one element

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125973

The previous fix was not general enough.

Fixes https://github.com/pytorch/pytorch/pull/125952

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -4020,10 +4020,10 @@ def index_select(x: TensorLike, dim: int, index: TensorLike):
     )
     if index.ndim == 0:
         index = index.unsqueeze(0)
-    if x.ndim == 0:
+    if x.numel() == 1:
         # Treat scalars as elements of \R^1
         # We cannot use x[idx] here as it accesses item() (??), hence this awkward construction
```
Comment (ezyang) [inline torch/_refs/__init__.py:4025]: Do we know why this happens?!

**Comment by ezyang** [comment]
This seems reasonable but I guess it doesn't work?

File: torch/_refs/__init__.py (hunk 1 above)
Comment (lezcano) [reply torch/_refs/__init__.py:4025]: See https://github.com/pytorch/pytorch/issues/105641

**Comment by lezcano** [comment]
The roll error is fine, but that pdist error I'll have to look into. I'll spin up an A100 box next week and make sure the fix I submit works lol

File: torch/_refs/__init__.py (hunk 1 above)
Comment (ezyang) [reply torch/_refs/__init__.py:4025]: Can we add an operator that is like `x[idx]` but instead produces a copy rather than view?

**Comment by ezyang** [comment]
This does not appear to fix it:

image

**Comment by ezyang** [comment]
And it's also difficult to tell how this could fix it, the trace appears to originate from C++ code

**Pull request closed without merging**